internal/xds/matcher/matcher_header.go

```diff
@@ -241,3 +241,34 @@ func (hcm *HeaderContainsMatcher) Match(md metadata.MD) bool {
 func (hcm *HeaderContainsMatcher) String() string {
 	return fmt.Sprintf("headerContains:%v%v", hcm.key, hcm.contains)
 }
+
+// HeaderStringMatcher matches on whether the header value matches against the
+// StringMatcher specified.
+type HeaderStringMatcher struct {
+	key           string
+	stringMatcher StringMatcher
+	invert        bool
+}
+
+// NewHeaderStringMatcher returns a new HeaderStringMatcher.
+func NewHeaderStringMatcher(key string, sm StringMatcher, invert bool) *HeaderStringMatcher {
+	return &HeaderStringMatcher{
+		key:           key,
+		stringMatcher: sm,
+		invert:        invert,
+	}
+}
+
+// Match returns whether the passed in HTTP Headers match according to the
+// specified StringMatcher.
+func (hsm *HeaderStringMatcher) Match(md metadata.MD) bool {
+	v, ok := mdValuesFromOutgoingCtx(md, hsm.key)
+	if !ok {
+		return false
+	}
+	return hsm.stringMatcher.Match(v) != hsm.invert
```

Comment:
The behavior here is unclear in some cases:

1. If a header is missing, should it always report `false`?  Or should it report `true` if `invert=true`?
2. If a header value is `""`, is that distinct from the header being missing, in any way?
3. Is the check immediately above designed to only detect an error in usage?  Because it doesn't consider (1) at least.

Reply:
See my chat thread. I think this is WAI.

Reply:
1. See Eric's explanation here: https://github.com/grpc/grpc-go/issues/4896
2. Yes, a set empty header string is logically distinct from it not being presence
3. No, this is part of the logic for the header matcher (not an error in usage, a header matcher can be called with metadata not containing the header key set in the matcher), as per explanation in 1